Refactor Snapcast client and group classes to use a common base clase

mill1000 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Refactor the `SnapcastGroupDevice` and `SnapcastClientDevice` classes to use a common `SnapcastBaseDevice` base class to reduce some boilerplate code between the two.

Additionally:
* ~Make offline clients unavailable.~
* Add proper type annotations.
* Refactor the platform to use a DataUpdateCoordinator.
* Refactor entity management to be within the platform.
* Relocate service call handling within the classes.

This PR is the result of splitting the original PR https://github.com/home-assistant/core/pull/117372/ as requested by @luar123 and @emontnemery 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: N/A
- This PR is related to issue: N/A, but original PR is here https://github.com/home-assistant/core/pull/117372
- Link to documentation pull request: N/A

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

mill1000 committed:
Refactor Snapcast client and group classes to use a common base class.

- Reorganize properties so they're closer to associated methods

mill1000 committed:
Fix typo in client

luar123 reviewed (approved):
LGTM

luar123 commented:
I am not sure about the Codecov CI error. There are no tests for this integration except for the configflow.

joostlek commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -114,195 +119,197 @@ class SnapcastGroupDevice(MediaPlayerEntity):
         self._entry_id = entry_id
-        self._attr_unique_id = f"{GROUP_PREFIX}{uid_part}_{self._group.identifier}"
+
+    def _append(self) -> None:
+        """Add self to appropriate list."""
+        raise NotImplementedError
+
+    def _remove(self) -> None:
+        """Remove self from appropriate list."""
+        raise NotImplementedError
+
+    @property
+    def current_group(self) -> Snapgroup:
+        """Return the group."""
+        raise NotImplementedError
```
So why do we add those?

mill1000 replied:
The `current_group` property allows a common definition for `source`, `source_list` and `async_select_source` in the base class. In retrospect, it should probably be renamed to `_current_group` as it is only accessed from within the classes.

Similarly `_append` and `_remove` allow common definitions for `async_added_to_hass` and `async_will_remove_from_hass` in the base class. I had felt that raising `NotImplementedError` was preferred to the default `AttributeError` in the case a developer forgets to implement the methods in the derived class.

joostlek replied:
Is there a way we can refactor this? We don't want to reference Entities directly after they're created so we should not store them in such list.

mill1000 replied:
Ok. What is the preferred way to reference an entity after it is created?

joostlek replied:
Not :)

Using entity services, which call a function on the entity (see snapshot and that other one). 

Other services usually directly reference the client and use that to send requests to the device or service, but i dont think that's a good fit here

mill1000 replied:
I see. The lists are primarily used to determine if a client/group entity already exists. Since clients/groups can come and go from the Snapcast server we need to be able to determine if an entity has already been created.

joostlek replied:
Where is the logic where we add an entity?

joostlek replied:
We can create a list of identifiers for which we created entities. Anytime we want to add one, we can check that list

joostlek replied:
Ideally we move the code that is required for adding entities to the platform and keep it there. Check AirGradient for an example (withings also is a good one, but it's bigger)

luar123 replied:
In the current implementation each entity adds itself to the corresponding list, e.g. `self.hass.data[DOMAIN][self._entry_id].groups` in `async_added_to_hass()`.
If there is an update from the server (only a single callback exists) this lists are used to find out which entities need to be added/removed.
> We can create a list of identifiers for which we created entities. Anytime we want to add one, we can check that list

This works for adding entities, but how can we remove entities based on the identifier?

joostlek replied:
Do the same thing, compare with your current list and if you have leftovers remove them. Checkout tedee for an example

mill1000 replied:
Ok, so it appears the refactor would probably be:
A. Refactor the `HomeAssistantSnapcast` class into a proper update coordinator derived from `DataUpdateCoordinator`.
B. Push entity addition/removal into `async_setup_entry` by registering a listener with the coordinator.

mill1000 replied:
@luar123 I think the removal would work by looking up the entities in the registry via their identifier.

mill1000 replied:
Ok, that ended up being a lot more work than I anticipated, but I think I've got it,

joostlek commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -83,29 +88,29 @@ async def async_setup_entry(
-async def handle_async_join(entity, service_call):
+async def handle_async_join(entity, service_call) -> None:
     """Handle the entity service join."""
     if not isinstance(entity, SnapcastClientDevice):
         raise TypeError("Entity is not a client. Can only join clients.")
     await entity.async_join(service_call.data[ATTR_MASTER])
 
 
-async def handle_async_unjoin(entity, service_call):
+async def handle_async_unjoin(entity, service_call) -> None:
     """Handle the entity service unjoin."""
     if not isinstance(entity, SnapcastClientDevice):
         raise TypeError("Entity is not a client. Can only unjoin clients.")
     await entity.async_unjoin()
 
 
-async def handle_set_latency(entity, service_call):
+async def handle_set_latency(entity, service_call) -> None:
     """Handle the entity service set_latency."""
     if not isinstance(entity, SnapcastClientDevice):
         raise TypeError("Latency can only be set for a Snapcast client.")
     await entity.async_set_latency(service_call.data[ATTR_LATENCY])
```
Why aren't these set as function inside of the media player object and then referenced like `snapshot` and `async_restore`?

mill1000 replied:
Good question. This was how there were originally defined and I did not change that. It was probably required in the old implementation because `SnapcastGroupDevice` didn't implement `async_set_latency` and had no base class it could fallback to.

mill1000 replied:
I moved the services into the classes but I think there's something wrong with the schema. When using developer tools to call the service the YAML generated by the UI is incorrect.

mill1000 replied:
Gotcha. Well my preference would be to leave them untouched, as I have changes ready to go that will implement grouping according to the [arch ](https://github.com/home-assistant/architecture/issues/364#issuecomment-625405413) that would make these services obsolete.

mill1000 committed:
Overhaul Snapcast to use data coordinator in push mode

- Use async_setup method to start server within coordinator
- Move entity creation/removal into platform
- Ensure SnapcastCoordinatorEntity's async_added_to_hass is called

mill1000 committed:
Make client unavailable if offline

mill1000 committed:
Fix typing for derived classes

mill1000 committed:
Move service handling into classes

mill1000 committed:
Put Derived CoordinatorEntity in 'entity' module per pylint

luar123 commented on homeassistant/components/snapcast/coordinator.py:
```diff
@@ -0,0 +1,75 @@
+        self._available = False
+
+        self._server.set_on_update_callback(self._on_update)
+        self._server.set_new_client_callback(self._on_update)
```
I guess this is fine. Adds some unnecessary overhead, but usually there won't be thousands of clients.

mill1000 replied:
What;s the performance concern? That adding a new client would cause unnecessary state updates for the others?

luar123 replied:
Yes, running on_update with looping through entities when just a client needs to be added is unnecessary. But I guess the overhead is negligible.

luar123 reviewed (commented):
If I will find the time I can test it, but so far looks ok to me (I don't know the details of coordinator implementation, though).

joostlek commented on homeassistant/components/snapcast/coordinator.py:
```diff
@@ -0,0 +1,75 @@
+        """Snapserver on_update callback."""
+        # Assume availability if an update is received.
+        self._available = True
+        self.async_set_updated_data(None)
```
Rather use
```suggestion
        self.async_update_listeners()
```

mill1000 replied:
Thanks for the feedback, didn't know I could update the listeners directly..

joostlek commented on homeassistant/components/snapcast/coordinator.py:
```diff
@@ -0,0 +1,75 @@
+        self._available = True
+        self.async_set_updated_data(None)
+
+    def _on_disconnect(self, ex):
+        """Snapsever on_disconnect callback."""
+        self._available = False
+        self.async_set_updated_data(None)
```
I am asuming that `ex` is an exception, you can use `self.async_set_update_error` to set it and automatically set all entities to unavailable

mill1000 replied:
Done.

joostlek commented on homeassistant/components/snapcast/coordinator.py:
```diff
@@ -0,0 +1,75 @@
+        self._server.set_new_client_callback(None)
+        self._server.stop()
+
+    @property
+    def available(self) -> bool:
+        """Get availability of the server."""
+        return self._available
```
I think we should replace all `self._available` with `self.last_update_success`

mill1000 replied:
Done.

joostlek commented on homeassistant/components/snapcast/entity.py:
```diff
@@ -0,0 +1,20 @@
+        """Create a Snapcast entity with an update coordinator."""
+        super().__init__(coordinator)
+
+    @property
+    def available(self) -> bool:
+        """Check device availability."""
+        return self.coordinator.available
```
(continued from last comment)

Then we can also remove this one, as the default implementation of available of a coordinator entity is the same

mill1000 replied:
Done.

joostlek commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -61,51 +77,111 @@ async def async_setup_entry(
+    def _check_entities() -> None:
+        nonlocal _known_group_ids, _known_client_ids
 
+        def _update_known_ids(known_ids, ids) -> tuple[str, str]:
```
typing

joostlek commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -61,51 +77,111 @@ async def async_setup_entry(
-    if not isinstance(entity, SnapcastClientDevice):
-        raise TypeError("Entity is not a client. Can only unjoin clients.")
-    await entity.async_unjoin()
+            # Update known IDs
+            known_ids -= ids_to_remove
+            known_ids |= ids_to_add
```
you can use `.update(..)`

mill1000 replied:
Done.

joostlek commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -61,51 +77,111 @@ async def async_setup_entry(
 
+            return ids_to_add, ids_to_remove
 
-async def handle_set_latency(entity, service_call):
-    """Handle the entity service set_latency."""
-    if not isinstance(entity, SnapcastClientDevice):
-        raise TypeError("Latency can only be set for a Snapcast client.")
-    await entity.async_set_latency(service_call.data[ATTR_LATENCY])
+        new_entities: list[SnapcastBaseDevice] = []
 
+        group_ids = {g.identifier for g in coordinator.server.groups}
+        groups_to_add, groups_to_remove = _update_known_ids(_known_group_ids, group_ids)
 
-class SnapcastGroupDevice(MediaPlayerEntity):
-    """Representation of a Snapcast group device."""
+        new_entities.extend(
+            [
+                SnapcastGroupDevice(
+                    coordinator, coordinator.server.group(group_id), host_id
+                )
+                for group_id in groups_to_add
+            ]
+        )
```
you can define `new_entities` the moment you have new entities

mill1000 replied:
Done.

joostlek commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -61,51 +77,111 @@ async def async_setup_entry(
+            if entity_id := entity_registry.async_get_entity_id(
+                MEDIA_PLAYER_DOMAIN,
+                DOMAIN,
+                SnapcastGroupDevice.get_unique_id(host_id, group_id),
+            ):
+                entity_registry.async_remove(entity_id)
+
+        for client_id in clients_to_remove:
+            if entity_id := entity_registry.async_get_entity_id(
+                MEDIA_PLAYER_DOMAIN,
+                DOMAIN,
+                SnapcastClientDevice.get_unique_id(host_id, client_id),
```
I think I prefer to have the `get_unique_id` functions to be outside of the class

mill1000 replied:
The unique_id format changes depending on if it's for a client or group, so not sure how I can move this out of the classes.

joostlek replied:
Oh but its fine to have 2 functions, I just don't think we should use a static method for it

mill1000 replied:
Hm...I'm not sure I follow. If I made them instance methods I wouldn't be able to use them in this context. I'd have to construct temporary objects and that seems wasteful.

joostlek replied:
No I mean, make them outside of the class. just have a `def _get_client_unique_id(.....)` and call that here and from the entity

mill1000 replied:
Ah gotcha.

mill1000 replied:
Ok, so a side effect of that change, I'd have to define constructors for both `SnapcastClientDevice` and `SnapcastGroupDevice` so they can call `_get_client_unique_id` and `_get_group_unique_id` respectively.

The classmethod approach allows the `SnapcastBaseDevice` constructor to handle everything.

It's not the end of the world but using the classmethod seems like the obvious choice, and I'm not sure what your concerns are with it.

joostlek replied:
Oh you don't have a constructor now, hmmmm. Then maybe this works what you have now.

joostlek commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -61,51 +77,111 @@ async def async_setup_entry(
+        client_ids = {c.identifier for c in coordinator.server.clients}
+        clients_to_add, clients_to_remove = _update_known_ids(
+            _known_client_ids, client_ids
+        )
```
After this point we know if we have new or to remove devices. We should rather quick return here if we have no action to do (unless this code is only executed when a player/group connects/disconnects)

mill1000 replied:
The code gets executed whenever there is a server update. @luar123 had some concerns about that in another comment too.

Implemented the early exit after the sets are updated.

joostlek commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -61,51 +77,111 @@ async def async_setup_entry(
+
+    coordinator.async_add_listener(_check_entities)
+
+    # Remove any existing entities
+    entity_registry = er.async_get(hass)
+    entity_registry.async_clear_config_entry(config_entry.entry_id)
```
euh why?

mill1000 replied:
While testing I found its possible to get the Snapserver and HA out of sync if a group/client disappears while HA is offline. (e.g. a group is removed while HA is restarting).

HA still has entities for these groups, but since their IDs aren't in `_known_group_ids` they are not removed on update.

Clearing the entities was simplest solution. Otherwise you'd have to rebuild the list of known IDs.

luar123 replied:
You are right, this is an issue with the current implementation, too. Could you read known ids from the registry instead?

mill1000 replied:
The client/group IDs could be parsed from the unique IDs if necessary. Is there any harm in starting from a blank state?

joostlek replied:
I'd rather not rely on that. What you can do is check all the devices you have after you forwarded the setups to the platforms, and compare them with all the devices you have for that config entry in the device registry

mill1000 replied:
> I'd rather not rely on that. What you can do is check all the devices you have after you forwarded the setups to the platforms

Sorry, I'm not following. At this point aren't I already in the platform setup?


> compare them with all the devices you have for that config entry in the device registry

I still have the problem that the entities in the device registry don't have the Snapcast group/client ID I need (unless parsed from the unique ID)

joostlek replied:
No I mean, this code can live in `__init__.py` and there you have a call where you forward the setups to the platforms. After that is completed, all devices are present. So if you then take all the devices you have in the device registry for this config entry, you can compare it with the data you have, and remove the ones you don't have data for.

But this is bigger than just a few lines, so I would recommend doing this in a separate PR

mill1000 replied:
Sorry, I might need this spelled out.

> No I mean, this code can live in __init__.py and there you have a call where you forward the setups to the platforms.

There was an earlier review comment that specifically wanted this code moved into the platform (i.e. `media_player.py`).

> So if you then take all the devices you have in the device registry for this config entry, you can compare it with the data you have, and remove the ones you don't have data for.

This would still require parsing the unique IDs to determine the group/client ID. I took your previous comments to indicate this would be undesirable.

joostlek replied:
> There was an earlier review comment that specifically wanted this code moved into the platform (i.e. media_player.py).

Could it be that this was about the creation?

> I took your previous comments to indicate this would be undesirable.

If you could link that I can maybe I can elaborate (or learn from that it can be recieved differently)

But I think if we remove it we can merge this and leave the device removal for a different PR discussion

mill1000 replied:
> Could it be that this was about the creation?

I suppose. :) Based on the examples it looked like the entire entity lifecycle was handled in the platform.



> If you could link that I can maybe I can elaborate (or learn from that it can be recieved differently)

This was the comment that I interpreted to mean "don't parse the unique IDs".
![image](https://github.com/user-attachments/assets/602c2ac2-6d36-4d51-aaab-07e7fd657b70)

https://github.com/home-assistant/core/pull/124499#discussion_r1820761762


> But I think if we remove it we can merge this and leave the device removal for a different PR discussion

Understood.

Thanks for your patience.

mill1000 replied:
Gotcha.

I've made the requested change for now.

joostlek commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -114,227 +190,232 @@ class SnapcastGroupDevice(MediaPlayerEntity):
+        host_id: str,
+    ) -> None:
+        """Initialize the base device."""
+        SnapcastCoordinatorEntity.__init__(self, coordinator)
```
```suggestion
        super().__init__(self, coordinator)
```

mill1000 replied:
Done.

joostlek commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -114,227 +190,232 @@ class SnapcastGroupDevice(MediaPlayerEntity):
-        """Set availability of group."""
-        self._attr_available = available
-        self.schedule_update_ha_state()
+    def __init__(
+        self,
+        coordinator: SnapcastUpdateCoordinator,
+        group: Snapgroup,
+        host_id: str,
+    ) -> None:
+        """Initialize the Snapcast group device."""
+        super().__init__(coordinator, group, host_id)
```
contructor can be removed

mill1000 replied:
Done.

joostlek commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -114,227 +190,232 @@ class SnapcastGroupDevice(MediaPlayerEntity):
+
+    _device: Snapclient
+
+    def __init__(
+        self,
+        coordinator: SnapcastUpdateCoordinator,
+        client: Snapclient,
+        host_id: str,
+    ) -> None:
+        """Initialize the Snapcast client device."""
+        super().__init__(coordinator, client, host_id)
```
idem

joostlek reviewed (approved):
I think this is going in a nice direction. At this point I did forget half the context, but we'll get there :)

joostlek reviewed (changes requested):
Wrong button

mill1000 committed:
Address review comments from @joostlek regarding _check_entities

- Fix typing of _update_known_ids method
- Exit early if not clients/groups to add/remove
- Use set() methods
- Construct entity list where its used

mill1000 committed:
Use data coordinators `last_update_success` and other methods to simplify availability

mill1000 committed:
Simplify constructors

mill1000 committed:
Catch OSError during setup and reraise as an UpdateError

mill1000 commented:
> At this point I did forget half the context, but we'll get there :)

Ha, yeah it went from "let me just add metadata real quick" to "well lets rewrite most the integration".

luar123 commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -114,227 +187,214 @@ class SnapcastGroupDevice(MediaPlayerEntity):
-        return self._client.muted
+    def available(self) -> bool:
+        """Check device availability."""
+        return super().available and self._device.connected
```
Please don't rely on the connected state for availability. Even if the device is turned off/disconnected the state can be read from the server and it can be controlled (e.q. you could change the source/volume/etc). Of course these changes take effect only when the device is connected again. So when the server is available all devices should be available, too. Disconnected devices then have the standby state: can be controlled, but are off.

mill1000 replied:
:+1:  Not my preference but understood.

luar123 reviewed (changes requested):
Run a short test in my environment. Works fine, except client device availability.

mill1000 committed:
Don't override client availability

mill1000 committed:
Don't blindly clear all entities during platform init per review

joostlek commented on homeassistant/components/snapcast/media_player.py:
```diff
@@ -114,227 +182,209 @@ class SnapcastGroupDevice(MediaPlayerEntity):
         self.async_write_ha_state()
 
+    @property
+    def volume_level(self) -> float | None:
```
```suggestion
    def volume_level(self) -> float:
```

joostlek reviewed (approved):
Okay this looks good. Also for a few followups.
1. Would it make sense to use built in joining and unjoining?
2. Can we make button entities of the snapshot/restore?
3. Can we make a number entity of the latency?

This would help us deprecate the services and make them easier in use :)

joostlek committed:
Update homeassistant/components/snapcast/media_player.py

mill1000 commented:
> Okay this looks good. Also for a few followups.
> 
>     1. Would it make sense to use built in joining and unjoining?

I have another branch where I have implemented this. I am aiming to have the metadata PR next and then follow up with the built-in join/unjoin.

>     2. Can we make button entities of the snapshot/restore?
>     3. Can we make a number entity of the latency?

Easy enough, but should be implement these in another PR?

joostlek commented:
Oh definitely all separate PRs

joostlek commented:
Oh wait I forgot to completely check. Forgot that part. Can we change the other entity lookup via the unique id?

mill1000 commented:
> Can we change the other entity lookup via the unique id?

Which lookup? The ones from this comment? https://github.com/home-assistant/core/pull/124499/files#r1816410375

luar123 commented:
> > Okay this looks good. Also for a few followups.
> > ```
> > 1. Would it make sense to use built in joining and unjoining?
> > ```
> 
> I have another branch where I have implemented this. I am aiming to have the metadata PR next and then follow up with the built-in join/unjoin.

Metadata support will be great :+1: and using native join/unjoin would also be a good idea. 

> 2. Can we make button entities of the snapshot/restore?

To be honest, I would not give more visibility to this functionality as it is not working properly. see #54609
Not sure if anyone uses it. In my opinion it is broken (or at least it is not working as one would expect) as for clients the stream and group is not captured. And groups are dynamic anyway. Would need to rework the python-snapcast package first.

> 3. Can we make a number entity of the latency?

You want to have a separate number entity for each client? Can we make them optional? I think we need to add device information to clients first, so it is clear that the client and number entities belong together.

More ideas for improvement (I never had time to look into these):
1. make groups optional (add option flow): #36442
2. stream sensor
3. change name service
4. Add control capabilities
5. Add tests
6. Add discovery flow

mill1000 commented:
> > Can we change the other entity lookup via the unique id?
> 
> Which lookup? The ones from this comment? https://github.com/home-assistant/core/pull/124499/files#r1816410375

@joostlek Could I bother you to clarify the above quoted question? Otherwise I think I'll mark this as ready again.

joostlek commented:
https://github.com/home-assistant/core/pull/124499/files#diff-26c6343a4873dacbc52a34d6a442b5e33edb7d1c3b1769667e6ecce90dba24caR367-R377

We should not lookup the entity from the entity component in hass.data, that's what I was referring to :)

(sorry for the late reaction)

mill1000 commented:
@joostlek Ok, that's fine, but how else do I go about getting the entities?

joostlek commented:
What exactly do you need? Because you can go from unique id to a lot more via the device, entity -registry and config entry

mill1000 commented:
Ultimately I need a reference to the entity object (i.e. SnapcastBaseDevice). I looked for example of using the entity registry or device registry but ultimately could only find an example using the entity component.

mill1000 commented:
>Ultimately I need a reference to the entity object (i.e. SnapcastBaseDevice). I looked for example of using the entity registry or device registry but ultimately could only find an example using the entity component.

@joostlek Sorry to bother you again. Can you provide any additional insight on my previous comment?

joostlek commented:
> Ultimately I need a reference to the entity object (i.e. SnapcastBaseDevice).

to do what?

mill1000 commented:
To be able to call methods and use properties of said object. e.g. To join a group, I need the Snapcast identifier of the master. This identifier isn't part of the device or entity registery unless I want to parse it from the unique ID.

joostlek commented:
That parsing is only splitting and taking the first item right? I think that's fine. Other option would be to fetch it through the config entries (I think)

mill1000 commented:
More or less. It's the last part of the unique ID that I'll need, so I can split on the underscore and take the last element (hopefully).

joostlek commented:
That would be a better idea yes

mill1000 committed:
Fetch Snapcast ID from entity's unique ID when joining

mill1000 committed:
Remove accidental logging

joostlek commented:
Sorry for merging so late, I remember I last looked at this in the bus and I probably got distracted midway. It was quite the journey but the result is awesome ❤️

mill1000 commented:
Great! Now I can make that metadata PR :)

joostlek commented:
It was literally that I saw a PR for Onkyo where I was like, wait, did I merge that snapcast one. If you feel like it, feel welcome to slide in my discord messages :)
